renderer.md

```diff
@@ -0,0 +1,28 @@
+## Disclaimer
+
+First of all this is an attempt to gather information about state of Renderer in dlang
+
+# What is a Renderer?
+ 
+Renderer contains a rendering state and provides functionality to draw primitives using specific color or texture. Renderer uses Image as a buffer to render to. Renderer is immediate one, i.e. it instantly modifies underlying Image during calling its methods. Renderer draw the following primitives:
```

Comment:
> it instantly modifies underlying Image during calling its methods

This can't be right? There's no computer made in the last 20 years that works that way...

Reply:
@TurkeyMan well, bad English is the language of the science
`it can immediately modify underlying Image as its method is called`

Reply:
Sure, there can be a `flush` method, but that's a deliberate request, and it's certainly not the case that "it instantly modifies underlying Image during calling its methods" as this text suggests.

Reply:
> it can immediately modify underlying Image as its method is called

That's still wrong.
How about: `the command is queued and flushed at some later time`

Reply:
`the command is queued and flushed at some later time`
isn't it more like `retained` not `immediate` one? 
I mean that if I call `immediateRenderer.drawLine(image, x1, y1, x2, y2)` after method returns the image will be modified.

Reply:
`flush` could be an optional primitive.
because for a 2D renderer that renders to PDF/Postscript etc you would need a `newPage` primitive that a Canvas renderer generally don't need.
These two operations don't make sense in the other context.

Reply:
> isn't it more like retained not immediate one?

No, that is how immediate rendering works.

> I mean that if I call immediateRenderer.drawLine(image, x1, y1, x2, y2) after method returns the image will be modified.

That sounds like something a js programmer might expect. It is wrong though, that's not now computers work. Rendering is a fundamentally asynchronous process.
We're D programmers, there are enough js programmers writing web-pages.

Reply:
Well, you're right, this is how it works. But shouldn't API hide these details under hood? If user called `drawLine` there is no way to cancel this operation even if it's not completed yet. In this sense it equals to statement `image has been modified`.

Reply:
> Well, you're right, this is how it works. But shouldn't API hide these details under hood?

No, core-api must be realistic. You can write something on top if you want to hide from reality.

> If user called drawLine there is no way to cancel this operation even if it's not completed yet. In this sense it equals to statement image has been modified.

It's not about cancellation, it's just that they need to wait until later, or block on a flush operation if they want to immediately access the results.

Reply:
I think a flush() operation to wait on pending results (or even enclosing draw operation behind begin/end paris) would be adequate and not too ominous.

Reply:
I think both may be necessary at the lowest-level command queuing api.

Reply:
both begin/end and flush? I can support that. begin/end can be reused for "printable" renderers (new page thing)